runtime/vm/LayoutFFITypeHelpers.cpp

```diff
@@ -100,32 +101,101 @@ LayoutFFITypeHelpers::getArrayFFIType(char **layout, UDATA nElements) {
+	if (0 == layoutStrFFITypeEntry->layoutStrLength) {
+		/* Malformed input. */
+		goto freeAllMemoryThenExit;
+	}
+	layoutStrFFITypeEntry->layoutStr = (U_8 *)j9mem_allocate_memory(layoutStrFFITypeEntry->layoutStrLength + 1, J9MEM_CATEGORY_VM_FFI);
+	if (NULL == layoutStrFFITypeEntry->layoutStr) {
+		goto freeAllMemoryThenExit;
+	}
+
+	memcpy(layoutStrFFITypeEntry->layoutStr, currentLayout, layoutStrFFITypeEntry->layoutStrLength);
+	layoutStrFFITypeEntry->layoutStr[layoutStrFFITypeEntry->layoutStrLength] = '\0';
+
+	/* Search the hashtable for the ffi_type if the same struct layout string exists. */
+	resultEntry = findLayoutStrFFIType(_vm->layoutStrFFITypeTable, layoutStrFFITypeEntry);
+	if (NULL != resultEntry) {
+		j9mem_free_memory(layoutStrFFITypeEntry->layoutStr);
+		j9mem_free_memory(layoutStrFFITypeEntry);
+		/* Return the duplicate ffi_type for the same struct if found in the hashtable. */
+		typeFFI = (ffi_type *)(resultEntry->structFFIType);
+		/* Reach the end of the nested struct layout string so as to move to
+		 * the next element in getStructFFITypeElements() if exists.
+		 */
+		(*layout) += resultEntry->layoutStrLength - 1;
+		goto done;
+	}
+
+	/* Skip over the '#' case. */
+	(*layout) = currentLayout + 1;
+
 	/* The struct layout is incremented for the '[' case in getStructFFITypeElements(), in which case
 	 * the remaining layout string is then traversed when getStructFFITypeElements() is called recursively.
 	 */
-	(*layout) += 1; // skip over the '#' case
-	ffi_type **structElements = getStructFFITypeElements(layout);
-	PORT_ACCESS_FROM_JAVAVM(_vm);
+	structElements = getStructFFITypeElements(layout);
+	if (NULL == structElements) {
+		goto freeAllMemoryThenExit;
+	}
+
 	typeFFI = (ffi_type *)j9mem_allocate_memory(sizeof(ffi_type), J9MEM_CATEGORY_VM_FFI);
 	if (NULL == typeFFI) {
-		setNativeOutOfMemoryError(_currentThread, 0, 0);
-		goto done;
+		goto freeAllMemoryThenExit;
 	}
 	typeFFI->size = 0;
 	typeFFI->alignment = 0;
 	typeFFI->type = FFI_TYPE_STRUCT;
 	typeFFI->elements = structElements;
 
+	/* Add the created ffi_type for the struct to the hashtable
+	 * if the corresponding layout string doesn't exist.
+	 */
+	layoutStrFFITypeEntry->structFFIType = (void *)typeFFI;
+	resultEntry = addLayoutStrFFIType(_vm->layoutStrFFITypeTable, layoutStrFFITypeEntry);
+	if (NULL == resultEntry) {
+		freeStructFFIType(typeFFI);
+		typeFFI = NULL;
+		goto freeAllMemoryThenExit;
+	}
+
 done:
 	return typeFFI;
+
+freeAllMemoryThenExit:
+	if (NULL != layoutStrFFITypeEntry) {
+		j9mem_free_memory(layoutStrFFITypeEntry->layoutStr);
+		j9mem_free_memory(layoutStrFFITypeEntry);
+	}
+	freeStructFFITypeElements(structElements);
+	setNativeOutOfMemoryError(_currentThread, 0, 0);
+	goto done;
 }
 
-ffi_type**
-LayoutFFITypeHelpers::getStructFFITypeElements(char **layout)
+ffi_type **
+LayoutFFITypeHelpers::getStructFFITypeElements(const char **layout)
 {
 	PORT_ACCESS_FROM_JAVAVM(_vm);
 
```

Comment:
Missing `const` on line 202.
The declaration of `elements` can move to where it gets a real value.
```c
	PORT_ACCESS_FROM_JAVAVM(_vm);
	const char *currentLayout = *layout;
	UDATA nElements = 0;
	UDATA elementCount = getIntFromLayout(&currentLayout);
	ffi_type **elements = (ffi_type **)j9mem_allocate_memory(sizeof(ffi_type *) * (elementCount + 1), J9MEM_CATEGORY_VM_FFI);
```

Reply:
Fixed.